core/src/listener/state.ts

```diff
@@ -177,33 +182,49 @@ export class State {
     await this.redisClient.set(this.stateName, JSON.stringify(updatedActiveListeners))
 
     const contractAddress = toAddListener.address
-    const observedBlockRedisKey = getObservedBlockRedisKey(contractAddress)
+    const contract = new ethers.Contract(contractAddress, this.abi, this.provider)
+    this.contracts = { ...this.contracts, [contractAddress]: contract }
+
     const latestBlock = await this.latestBlockNumber()
+    const observedBlock = await getObservedBlock({ service: this.service })
 
-    switch (this.listenerInitType) {
-      case 'clear':
-        // Clear metadata about previously observed blocks for a specific
-        // `contractAddress`.
-        await this.redisClient.set(observedBlockRedisKey, latestBlock - 1)
-        break
-
-      case 'latest':
-        await this.setObservedBlockNumberIfNotDefined(observedBlockRedisKey, latestBlock - 1)
-        break
-
-      default:
-        // [block number] initialization
-        await this.setObservedBlockNumberIfNotDefined(observedBlockRedisKey, latestBlock - 1)
-        for (let blockNumber = this.listenerInitType; blockNumber < latestBlock; ++blockNumber) {
-          const historyOutData: IHistoryListenerJob = {
-            contractAddress,
-            blockNumber
+    // fetch all unprocessed blocks and pass to the history queue
+    const unprocessedBlocks = await getUnprocessedBlocks({ service: this.service })
+    for (const block of unprocessedBlocks) {
+      await this.addBlockToHistoryQueue(contractAddress, block.blockNumber)
+    }
+
+    // if there is no observedBlock record in the db,
+    // use the listenerInitType to determine how to initialize the listener
+    if (observedBlock.service === '') {
+      switch (this.listenerInitType) {
+        case 'clear':
+          // Clear metadata about previously observed blocks for a specific
+          // `contractAddress`.
+          await upsertObservedBlock({ service: this.service, blockNumber: latestBlock - 1 })
+          break
+
+        case 'latest':
+          await upsertObservedBlock({ service: this.service, blockNumber: latestBlock - 1 })
+          break
```

Comment:
since clear and latest having the same logic how about merging it together as followed
```
        case 'clear':
        case 'latest':
          await upsertObservedBlock({ service: this.service, blockNumber: latestBlock - 1 })
          break
```